Using the Iris data set, do k-fold cross validation for Random Forest, SVM, and Logisitc Regression. Which one perform the best?
What happens if you change the Kernel? I.e, do you get 98%? If you use SVC (kernel = poly)

In [141]:
# load models from sklearn and iris data set
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
import numpy as np
from sklearn.datasets import load_iris

iris = load_iris()
dir(iris)

['DESCR',
 'data',
 'data_module',
 'feature_names',
 'filename',
 'frame',
 'target',
 'target_names']

In [142]:
# split the data 
from sklearn.model_selection import train_test_split

X = iris.data
y = iris.target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

In [143]:
# logistic regression 
lr = LogisticRegression()
lr.fit(X_train, y_train)
lr.score(X_test, y_test)

/Users/dani/Desktop/csc180/.venv/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.9777777777777777

In [144]:
# SVC
svm = SVC()
svm.fit(X_train, y_train)
svm.score(X_test, y_test)

0.9555555555555556

In [145]:
# random forest 
rf = RandomForestClassifier()
rf.fit(X_train, y_train)
rf.score(X_test, y_test)

0.9555555555555556

sample distribution is not fixed and model accuracy changes every time we re-run the split data 

In [146]:
# k-fold 
# from sklearn.model_selection import KFold

# kf = KFold(n_splits=3)
# kf

# stratisfied k-fold ensures that each fold preserves the percentage of samples for each class
from sklearn.model_selection import StratifiedKFold
kf = StratifiedKFold(n_splits=3, shuffle=True, random_state=42)

In [147]:
# visualize how kfold split works 
# for train_index, test_index in kf.split([0,1,2,3,4,5,6,7,8]):
#     print(train_index, test_index)

In [151]:
from sklearn.model_selection import cross_val_score

# define models
models = {
    "Random Forest": RandomForestClassifier(),
    "SVM (RBF)": SVC(kernel='rbf'),
    "Logistic Regression": LogisticRegression()
}

# evaluate each model
for name, model in models.items():
    scores = cross_val_score(model, X, y, cv=kf)

    print(f"{name}: Model Score Average = {np.mean(scores):.4f}")
    print(scores)

# different kernel
poly_svm = SVC(kernel='poly')
poly_scores = cross_val_score(poly_svm, X, y, cv=kf)
print(f"SVM (poly kernel): Model Score Average= {np.mean(poly_scores):.4f}")
print(scores)


Random Forest: Model Score Average = 0.9533
[0.98 0.94 0.94]
SVM (RBF): Model Score Average = 0.9667
[1.   0.94 0.96]
Logistic Regression: Model Score Average = 0.9667
[1.   0.94 0.96]
SVM (poly kernel): Model Score Average= 0.9667
[1.   0.94 0.96]


As we can see, the average accuracy for Random Forest is 0.95, 0.97 for SVM, and 0.97 for Logistic Regression. Changing the SVC kernel didn't make much of a difference here, as they both have the same model scores. All models performed very well with similar scores.